### PR TITLE
Improve publisher load time for 120k nodes to <1s

### LINKED ISSUE #912: Opcpublisher stops sending, after running for few days. Requires restart to resume telemetry
**Describe the bug**
Opcpublisher stops sending data to cloud, after running for few days. In the log, I can see 'msg send failures' keeps increasing on every send interval. After I restart opcpublisher, data starts to flow again.

**To Reproduce**
This issue we have seen in multiple VMs. When configured high amount of tags (above 10K), and run continuously for more than 2 weeks, message send failure occurs. From there on, no messages are received on the cloud. Please refer below log. 
The telemetry resumes after opcpublisher restart.

**Screenshots**
==========================================================================
[07:02:00 INF] OpcPublisher status @ 11/16/2020 07:02:00 (started @ 10/28/2020 09:50:32)
[07:02:00 INF] ---------------------------------
[07:02:00 INF] OPC sessions (configured/connected): 1/1
[07:02:00 INF] OPC subscriptions (configured/connected): 1/1
[07:02:00 INF] OPC monitored items (configured/monitored/to remove): 49445/49445/0
[07:02:00 INF] ---------------------------------
[07:02:00 INF] monitored items queue bounded capacity: 24000
[07:02:00 INF] monitored items queue current items: 0
[07:02:00 INF] monitored item notifications enqueued: 425815601
[07:02:00 INF] monitored item notifications enqueue failure: 13881
[07:02:00 INF] ---------------------------------
[07:02:00 INF] messages sent to IoTHub: 666352
[07:02:00 INF] last successful msg sent @: 11/15/2020 02:24:35
[07:02:00 INF] bytes sent to IoTHub: 136843849560
[07:02:00 INF] avg msg size: 205362
[07:02:00 INF] msg send failures: 37316
[07:02:00 INF] messages too large to sent to IoTHub: 0
[07:02:00 INF] times we missed send interval: 1619613
[07:02:00 INF] number of events: 425801720
[07:02:00 INF] ---------------------------------
[07:02:00 INF] current working set in MB: 346
[07:02:00 INF] --si setting: 10
[07:02:00 INF] --ms setting: 225000
[07:02:00 INF] --ih setting: Amqp_Tcp_Only
[07:02:00 INF] ==========================================================================

**Desktop (please complete the following information):**
 - OS: Ubuntu 
 - Version: 18.04LTS
 - Opcpublisher : v2.5.3

### PR DESCRIPTION
**Pull request by mregen**
- use IEqualityComparer for Distinct. reduce startup time for 120k nodes from 20 minutes to <1s

### CODE REVIEW

**Comment by karok2m** [comment]
@mregen, please link the GitHub issue to this PR.

**Comment by mregen** [comment]
This PR addresses a comment in the Issue #912: https://github.com/Azure/Industrial-IoT/issues/912#issuecomment-877170835

**Comment by mregen** [comment]
> @mregen, please link the GitHub issue to this PR.

only a comment in #912 is fixed...otherwise i believe it remains open.

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Storage/PublishedNodesJobConverter.cs [hunk 1]
```diff
@@ -71,130 +71,177 @@ public class PublishedNodesJobConverter {
...
+                                    NetworkMessageContentMask.DataSetMessageHeader
                         }
-                    }).ToList();
+                    }
+                });
+                _logger.Information("Created {jobs} jobs in {elapsed}", result.Count(), sw.Elapsed);
+
+                var counter = 0;
+                foreach (var job in result) {
+                    _logger.Verbose("groupId: {group}", job.WriterGroup.WriterGroupId);
```
Comment (karok2m) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Storage/PublishedNodesJobConverter.cs]: @mregen, can you please add some more information to the log message. It is not clear how this log can be used. The same comment for the log bellow on line `215`.

Comment (mregen) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Storage/PublishedNodesJobConverter.cs]: what kind of information do you mean? The output is pretty obvious:
```
[11:22:55 INF Microsoft.Azure.IIoT.OpcUa.Edge.Publisher.Models.PublishedNodesJobConverter] Read 1 items from published nodes file in "00:00:00.3329831" 
[11:22:55 INF Microsoft.Azure.IIoT.OpcUa.Edge.Publisher.Models.PublishedNodesJobConverter] Created 1 groups in "00:00:00.0024429" 
[11:22:55 INF Microsoft.Azure.IIoT.OpcUa.Edge.Publisher.Models.PublishedNodesJobConverter] Created 1 flattened groups in "00:00:00.1736669" 
[11:22:55 INF Microsoft.Azure.IIoT.OpcUa.Edge.Publisher.Models.PublishedNodesJobConverter] Created 1 jobs in "00:00:00.0672013" 
[11:22:55 INF Microsoft.Azure.IIoT.OpcUa.Edge.Publisher.Models.PublishedNodesJobConverter] Total count of OpcNodes after job conversion: 125012 
[11:22:55 INF Microsoft.Azure.IIoT.OpcUa.Edge.Publisher.Models.PublishedNodesJobConverter] Converted items to jobs in "00:00:00.3296201" 
```

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Storage/PublishedNodesJobConverter.cs [hunk 2]
```diff
@@ -71,130 +71,177 @@ public class PublishedNodesJobConverter {
...
-                                        NetworkMessageContentMask.SequenceNumber |
-                                        NetworkMessageContentMask.PayloadHeader |
-                                        NetworkMessageContentMask.Timestamp |
-                                        NetworkMessageContentMask.DataSetClassId |
-                                        NetworkMessageContentMask.NetworkMessageHeader |
-                                        NetworkMessageContentMask.DataSetMessageHeader
+                        }
+                    ).ToList()
+                ).ToList();
+                _logger.Information("Created {n} flattened groups in {elapsed}", flattenedGroups.Count, sw.Elapsed);
```
Comment (koepalex) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Storage/PublishedNodesJobConverter.cs]: This log isn't so relevant for the customer, please change to Debug log level

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Storage/PublishedNodesJobConverter.cs [hunk 3]
```diff
@@ -71,130 +71,177 @@ public class PublishedNodesJobConverter {
...
-                            .Distinct((a, b) => a.Id == b.Id && a.DisplayName == b.DisplayName && a.DataSetFieldId == b.DataSetFieldId &&
-                                        a.OpcSamplingInterval == b.OpcSamplingInterval)
-                            .Batch(1000))
-                        .Select(opcNodes => new PublishedDataSetSourceModel {
+                    // Select and batch nodes into published data set sources
+                    item => GetNodeModels(item, legacyCliModel.ScaleTestCount.GetValueOrDefault(1)),
+                    // Comparer for connection information
+                    new FuncCompare<ConnectionModel>((x, y) => x.IsSameAs(y))
+                ).ToList();
+                _logger.Information("Created {n} groups in {elapsed}", group.Count, sw.Elapsed);
```
Comment (koepalex) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Storage/PublishedNodesJobConverter.cs]: >i [](http://example.com/codeflow?start=32&length=1)

This log isn't so relevant for the customer, please change to Debug log level

Comment (mregen) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Storage/PublishedNodesJobConverter.cs]: removed

**Comment by koepalex** [comment]
@mregen could you address the LGTM issue, then we are good to merge this PR

**Review by karok2m — DISMISSED** [review]

**Comment by mregen** [comment]
> @mregen could you address the LGTM issue, then we are good to merge this PR

LGTM is really great to catch this! :shipit:

**Review by karok2m — APPROVED** [review]

**Pull request merged**